ddtrace/bootstrap/sitecustomize.py

```diff
@@ -2,30 +2,105 @@
 """
-import logging
-import os
 import sys
-from typing import Any
-from typing import Dict
 
 
-# Perform gevent patching as early as possible in the application before
-# importing more of the library internals.
-if os.environ.get("DD_GEVENT_PATCH_ALL", "false").lower() in ("true", "1"):
-    import gevent.monkey
+MODULES_LOADED_AT_STARTUP = frozenset(sys.modules.keys())
+MODULES_THAT_TRIGGER_CLEANUP_WHEN_INSTALLED = ("gevent",)
 
-    gevent.monkey.patch_all()
 
+import os  # noqa
+
+
+MODULES_TO_NOT_CLEANUP = {"atexit", "asyncio", "attr", "concurrent", "ddtrace", "logging"}
+if sys.version_info <= (2, 7):
+    MODULES_TO_NOT_CLEANUP |= {"encodings", "codecs"}
+    import imp
+
+    _unloaded_modules = []
+
+    def is_installed(module_name):
+        try:
+            imp.find_module(module_name)
+        except ImportError:
+            return False
+        return True
+
+
+else:
+    import importlib
+
+    def is_installed(module_name):
+        return importlib.util.find_spec(module_name)
+
+
+def should_cleanup_loaded_modules():
+    dd_unload_sitecustomize_modules = os.getenv("DD_UNLOAD_MODULES_FROM_SITECUSTOMIZE", default="0").lower()
+    if dd_unload_sitecustomize_modules not in ("1", "auto"):
+        return False
+    elif dd_unload_sitecustomize_modules == "auto" and not any(
+        is_installed(module_name) for module_name in MODULES_THAT_TRIGGER_CLEANUP_WHEN_INSTALLED
+    ):
+        return False
+    return True
+
+
+def cleanup_loaded_modules(aggressive=False):
+    """
+    "Aggressive" here means "cleanup absolutely every module that has been loaded since startup".
+    Non-aggressive cleanup entails leaving untouched certain modules
+    This distinction is necessary because this function is used both to prepare for gevent monkeypatching
+    (requiring aggressive cleanup) and to implement "module cloning" (requiring non-aggressive cleanup)
+    """
+    modules_loaded_since_startup = set(_ for _ in sys.modules if _ not in MODULES_LOADED_AT_STARTUP)
+    # Figuring out modules_loaded_since_startup is necessary because sys.modules has more in it than just what's in
+    # import statements in this file, and unloading some of them can break the interpreter.
+    modules_to_cleanup = modules_loaded_since_startup - MODULES_TO_NOT_CLEANUP
+    if aggressive:
+        modules_to_cleanup = modules_loaded_since_startup
+    # Unload all the modules that we have imported, except for ddtrace and a few
+    # others that don't like being cloned.
+    # Doing so will allow ddtrace to continue using its local references to modules unpatched by
+    # gevent, while avoiding conflicts with user-application code potentially running
+    # `gevent.monkey.patch_all()` and thus gevent-patched versions of the same modules.
+    for m in modules_to_cleanup:
+        if not aggressive and any(
+            m.startswith("%s." % module_to_not_cleanup) for module_to_not_cleanup in MODULES_TO_NOT_CLEANUP
+        ):
```

Comment:
I feel like I had to jump back and forth a bit to understand exactly when/what modules would be unloaded, I wonder if writing it like this would be easier to follow?


```python
# We cannot iterate over a dict which might change in size...
#   do we have to cast the keys call to something else to avoid this?
for module_name in set(sys.modules.keys()):
    # Modules to never unload (this wasn't in original code, just showcasing concept)
    if module_name in ("sys", ):
        continue

    # If we are being aggressive, unload all modules
    if aggressive:
        del sys.modules[module_name]
        continue
    
    # I don't like how I formatted this, and I think I got the logic wrong, but hopefully you get the idea
    should_cleanup_module = (
        module_name not in MODULES_TO_NOT_CLEANUP
        or
        not any(module_name.startswith("%s." % module_to_not_cleanup) for module_to_not_cleanup in MODULES_TO_NOT_CLEANUP)
    )
    if should_cleanup_module:
        del sys.modules[module_name]
```


another idea that came to me.

```python
for module_name in set(sys.modules.keys()):
    if aggressive:
        del sys.modules[module_name]
        continue

    for module_to_not_cleanup in MODULES_TO_NOT_CLEANUP:
        if module_name == module_to_not_cleanup:
            break
        elif module_name.startswith("%s." % module_to_not_cleanup):
            break
   else:    
       del sys.modules[module_name]
```


no changes really needed, feel free to resolve/ignore this comment. these are two ideas that came to me that I thought might make this function easier to follow, but I understand personal preferences are mostly likely biasing my suggestions.